feat(AWS Lambda): Add support for self-managed `kafka` event

lewgordon opened the pull request:
<!-- ⚠️⚠️ Acknowledge ALL below remarks -->
<!-- ⚠️⚠️ PR will not be processed if it doesn't meet outlined criteria -->

<!-- ⚠️⚠️ Do not propose PR's without prior agreement on a solution in the corresponding issue -->
<!-- ⚠️⚠️ Only documentation updates and obvious bug fixes are welcome without it -->

<!--
⚠️⚠️ Ensure to follow code style guidelines
https://github.com/serverless/serverless/blob/master/CONTRIBUTING.md#code-style
-->

<!--
⚠️⚠️ Ensure to cover changes with tests written according to test guidelines
https://github.com/serverless/serverless/blob/master/test/README.md
-->

<!-- ⚠️⚠️ Ensure that support for Node.js v10 is maintained. -->

<!--
⚠️⚠️ Ensure that the proposed change passes CI. Confirm that by running the following scripts:
• npm run prettier-check
• npm run lint
• npm test
-->

<!--
⚠️⚠️ If proposed change touches integration with AWS services, confirm integration tests pass:
https://github.com/serverless/serverless/blob/master/test/README.md#aws-integration-tests
-->

<!-- ⚠️⚠️ After your PR is submitted, review the final CI status and address eventual failure -->

<!-- ⚠️⚠️ Answer below questions -->

<!--
Q1: Provide a link to the corresponding issue

• If PR *partially* addresses issue, ensure to rename "Closes" to "Addresses" ("Closes" term will automatically close an issue on PR merge)
• If it's a documentation update or obvious bug fix that has no corresponding issue, replace this line with a short description of made changes
-->

Closes: #8718

lewgordon commented:
I have an extra commit for documentation, but for some reason it isn't showing up here! :shrug:

pgrzesik commented on lib/plugins/aws/lib/naming.js:
```diff
@@ -422,6 +422,14 @@ module.exports = {
+  // Kafka
+  getKafkaEventLogicalId(functionName, topicName) {
+    const normalizedFunctionName = this.getNormalizedFunctionName(functionName);
+    // Both topicName is trimmed to 158 chars to avoid going over 255 character limit
```
I guess "both" is not needed here as only `topicName` is trimmed

pgrzesik commented on lib/plugins/aws/package/compile/events/msk/index.js:
```diff
@@ -40,6 +40,79 @@ class AwsCompileMSKEvents {
+    const secretsManagerArnRegex =
+      'arn:[a-z-]+:secretsmanager:[a-z0-9-]+:\\d+:secret:[A-Za-z0-9/_+=.@-]+';
+
+    this.serverless.configSchemaHandler.defineFunctionEvent('aws', 'kafka', {
```
I believe in this case it would be better to go with separate module/class responsible for `kafka` event compilation, as in current form, not much is shared between the two and both filename and class here suggest that it's only responsible for `msk` event. What do you think? Do you think there's a reason to keep them both in this module?

lewgordon replied:
Yeah I think that's reasonable. I think at first I kept them in the same file since they were mostly related. But I think it makes sense to separate them.

pgrzesik commented on lib/plugins/aws/package/compile/events/msk/index.js:
```diff
@@ -40,6 +40,79 @@ class AwsCompileMSKEvents {
+                    type: 'string',
+                    pattern: 'subnet-[a-z0-9]+',
+                  },
+                  { type: 'object' },
```
Just to confirm, this is here to support potential support for intrinsic functions with mandatory `!Sub` to construct proper value, correct?

lewgordon replied:
Yes, exactly.

pgrzesik commented on test/unit/lib/plugins/aws/package/compile/events/msk/index.test.js:
```diff
@@ -9,11 +9,434 @@ chai.use(require('chai-as-promised'));
+          };
+        },
+      },
+    ].forEach((eventConfig) => {
```
This might be my personal prefrence, but did you consider e.g. abstracting away `it` block with test to a function and instead of iterating here, do something like

```
const runCompileEventSourceMappingTest = (eventConfig) => {...}

it('for event xxx', async () => {
const eventConfig = {...}
await runCompileEventSourceMapppingTest(eventConfig)
})
```

The result is similar, but each separate test is a bit easier to grasp (as mentioned, it might be my personal preference and it might not be "easier to grasp" for everyone).

lewgordon replied:
I hadn't thought about that, but I do like it.  At least for what I have here, since the payloads are a bit unwieldy.

pgrzesik reviewed (changes requested):
Thank you @lewgordon :+1: Overall it's a great work, I have a few comments, please check them out and let me know what you think. I'll also ask @medikoo for an extra review to asses the chosen approach as correct.

pgrzesik commented on lib/plugins/aws/package/compile/events/kafka.js:
```diff
@@ -0,0 +1,226 @@
+      const functionObj = this.serverless.service.getFunction(functionName);
+      const cfTemplate = this.serverless.service.provider.compiledCloudFormationTemplate;
+
+      // It is required to add the following statement in order to be able to connect to MSK cluster
```
Just to confirm, is it still required at all times or only when VPC-based `accessConfiguration` is used?

lewgordon replied:
Hmm, not sure, but will double check.

lewgordon replied:
I believe it's only required for VPC-based `accessConfiguration`.

pgrzesik commented on docs/providers/aws/events/kafka.md:
```diff
@@ -0,0 +1,104 @@
+    handler: handler.compute
+    events:
+      - kafka:
+          accessConfigurations:
```
Which approach is the valid one in the end? I see in the documentation here that `accessConfigurations` are supposed to be an Object, however, in tests and in processing logic it expects it to be an array. Based on recent discussion here https://github.com/serverless/serverless/issues/8718 I got the impression that we could go with object-based approach.

lewgordon replied:
Yep, good catch.  I think my documentation is right, but the code may need some fixes.

pgrzesik replied:
Thanks for clarification :raised_hands:

pgrzesik commented on lib/plugins/aws/package/compile/events/kafka.js:
```diff
@@ -0,0 +1,226 @@
+      type: 'object',
+      properties: {
+        accessConfigurations: {
+          type: 'array',
```
Should be an array or an object?

lewgordon replied:
An object.

pgrzesik commented on test/unit/lib/plugins/aws/package/compile/events/kafka.test.js:
```diff
@@ -0,0 +1,453 @@
+
+chai.use(require('chai-as-promised'));
+
+describe('AwsCompileKafkaEvents', () => {
```
I'm not sure if this is documented, but it's recommended to name new describe blocks after the name of the test file, so in this case it will be `test/unit/lib/plugins/aws/package/compile/events/kafka.test.js`

pgrzesik reviewed (changes requested):
Thank you @lewgordon, in general it looks good, I have a few questions though, please see my comments and let me know what you think

lewgordon commented:
I also just verified that `BASIC_AUTH` is not supported for self-managed Apache Kafka, so I removed it.

pgrzesik commented on lib/plugins/aws/package/compile/events/kafka.js:
```diff
@@ -0,0 +1,200 @@
+          type: 'boolean',
+        },
+        bootstrapServers: {
+          type: 'array',
```
Should we also specify the type of item that's allowed here?

pgrzesik commented on lib/plugins/aws/package/compile/events/kafka.js:
```diff
@@ -0,0 +1,200 @@
+        if (secretsManagerStatement.Resource.length) {
+          statement.push(secretsManagerStatement);
+        }
+        if (needsEc2Permissions) {
```
:+1:

pgrzesik commented on test/unit/lib/plugins/aws/package/compile/events/kafka.test.js:
```diff
@@ -0,0 +1,534 @@
+      expect(eventSourceMappingResource.Properties).to.deep.equal(eventConfig.resource(awsNaming));
+    };
+
+    const requirePropertyTest = async (event) => {
```
Our standard approach is not to test schema validation in such manner - we instead test all supported scenarios which you already have. I believe these tests could be removed

pgrzesik reviewed (changes requested):
Looks good overall, I have two minor comments, please let me know what you think. I'd also like to wait with final merge until @medikoo has a chance to re-review it :+1:

medikoo commented on lib/plugins/aws/package/compile/events/kafka.js:
```diff
@@ -0,0 +1,200 @@
+          type: 'object',
+          minProperties: 1,
+          properties: {
+            vpcSubnet: arrayOrStringSchema('subnet-[a-z0-9]+'),
```
We don't need to use `arrayOrStringSchema`, as in schema validator we have _to array_ coercion configured to be _on_ (https://ajv.js.org/docs/coercion.html)

So we should simply configure those properties as arrays.  Singular values will automatically be allowed, and in code we can unconditionally expect arrays (as coercion will ensure all values are converted to arrays during validation phase)

pgrzesik replied:
Thanks for bringing that up @medikoo - I was not aware of that :+1:

lewgordon replied:
@medikoo after changing this, should I inline the lambda function or keep as is?  I wasn't sure what was more readable.

medikoo replied:
I would inline the schema (when inspecting the schema, then it's easier to acknowledge what's defined)

medikoo commented on lib/plugins/aws/package/compile/events/kafka.js:
```diff
@@ -0,0 +1,200 @@
+            saslScram256Auth: arrayOrStringSchema(secretsManagerArnRegex),
+            saslScram512Auth: arrayOrStringSchema(secretsManagerArnRegex),
+          },
+        },
```
Misses `additionalProperties: false`

medikoo commented on lib/plugins/aws/package/compile/events/kafka.js:
```diff
@@ -0,0 +1,200 @@
+      let needsEc2Permissions = false;
+
+      functionObj.events.forEach((event) => {
+        if (event.kafka == null) {
```
Let's improve it to simply `if (!event.kafka) return;`

medikoo commented on lib/plugins/aws/package/compile/events/kafka.js:
```diff
@@ -0,0 +1,200 @@
+        kafkaResource.Properties.SourceAccessConfigurations = [];
+        Object.entries(event.kafka.accessConfigurations).forEach(
+          ([accessConfigurationType, value]) => {
+            const accessConfigurationValues = typeof value === 'string' ? [value] : value;
```
After we simplify schema, we can unconditionally expect an _array_

medikoo reviewed (changes requested):
@lewgordon that's outstanding! Thank you.

In addition to @pgrzesik comments, I have just few minor improvement suggestions from me.

Let me know if anything is not clear

lewgordon commented:
In addition to the requested changes in the PR, I made a few changes to the documentation that I had forgot about.  I removed references to using `basicAuth`, as it is not valid for this event type.  I also removed an example that uses an `accessConfiguration` that is not in the schema.

medikoo commented on lib/plugins/aws/package/compile/events/kafka.js:
```diff
@@ -0,0 +1,212 @@
+                pattern: secretsManagerArnRegex,
+              },
+            },
+            additionalProperties: false,
```
This should be set in same level as `properties` (currently seems as an extra property definition)

medikoo commented on lib/plugins/aws/package/compile/events/kafka.js:
```diff
@@ -0,0 +1,212 @@
+'use strict';
+class AwsCompileKafkaEvents {
```
Let's add a line space between `'use strict'` and code (sorry we should have a lint rule for that)

medikoo commented on test/unit/lib/plugins/aws/package/compile/events/kafka.test.js:
```diff
@@ -0,0 +1,534 @@
+      ).to.be.eventually.rejectedWith(ServerlessError);
+    };
+
+    it('should require access configurations', async () => {
```
As @pgrzesik pointed, let's not introduce tests that confirm that schema validation errors are thrown on invalid values

We have some sanity tests for it in other place, and otherwise schema validation engine is extensively tested in its own repository.

medikoo reviewed (changes requested):
Thank you @lewgordon ! Looks great, I have just few comments

medikoo reviewed (approved):
Thank you @lewgordon Looks great to me!

I'll leave it for final look to @pgrzesik

lewgordon commented:
Thank you for your help and patience @medikoo and @pgrzesik !

pgrzesik reviewed (approved):
Thank you @lewgordon, outstanding work :+1:
